lighthouse-core/config/budget.js

```diff
@@ -39,46 +66,49 @@ class Budget {
       'other',
       'third-party',
     ];
-    if (!validResourceTypes.includes(resourceBudget.resourceType)) {
-      throw new Error(`Invalid resource type: ${resourceBudget.resourceType}. \n` +
+    // Assume resourceType is an allowed string, throw if not.
+    if (!validResourceTypes.includes(/** @type {LH.Budget.ResourceType} */ (resourceType))) {
+      throw new Error(`Invalid resource type: ${resourceType}. \n` +
         `Valid resource types are: ${ validResourceTypes.join(', ') }`);
     }
-    if (isNaN(resourceBudget.budget)) {
-      throw new Error('Invalid budget: ${resourceBudget.budget}');
+    if (!isNumber(budget)) {
+      throw new Error(`Invalid budget: ${budget}`);
     }
     return {
-      resourceType,
+      resourceType: /** @type {LH.Budget.ResourceType} */ (resourceType),
       budget,
     };
   }
 
   /**
-   * @param {LH.Budget.TimingBudget} timingBudget
+   * @param {Record<string, unknown>} timingBudget
    * @return {LH.Budget.TimingBudget}
    */
   static validateTimingBudget(timingBudget) {
     const {metric, budget, tolerance, ...invalidRest} = timingBudget;
     Budget.assertNoExcessProperties(invalidRest, 'Timing Budget');
 
+    /** @type {Array<LH.Budget.TimingMetric>} */
     const validTimingMetrics = [
       'first-contentful-paint',
       'first-cpu-idle',
       'interactive',
       'first-meaningful-paint',
       'estimated-input-latency',
     ];
-    if (!validTimingMetrics.includes(timingBudget.metric)) {
-      throw new Error(`Invalid timing metric: ${timingBudget.metric}. \n` +
+    // Assume metric is an allowed string, throw if not.
+    if (!validTimingMetrics.includes(/** @type {LH.Budget.TimingMetric} */ (metric))) {
+      throw new Error(`Invalid timing metric: ${metric}. \n` +
         `Valid timing metrics are: ${validTimingMetrics.join(', ')}`);
     }
-    if (isNaN(timingBudget.budget)) {
-      throw new Error('Invalid budget: ${timingBudget.budget}');
+    if (!isNumber(budget)) {
+      throw new Error(`Invalid budget: ${budget}`);
     }
-    if (timingBudget.tolerance !== undefined && isNaN(timingBudget.tolerance)) {
-      throw new Error('Invalid tolerance: ${timingBudget.tolerance}');
+    if (typeof tolerance !== 'undefined' && !isNumber(tolerance)) {
```

Comment:
I thought you were against `typeof x === 'undefined'` checks @brendankenny ? 😛

Reply:
> I thought you were against `typeof x === 'undefined'` checks @brendankenny ? 😛

haha, I am! For some reason typescript won't treat it as `number | undefined` with `tolerance !== undefined`, and it's really annoying :)